flink/src/main/java/org/apache/iceberg/flink/FlinkSchemaUtil.java

```diff
@@ -98,4 +102,22 @@ public static TableSchema toSchema(RowType rowType) {
     }
     return builder.build();
   }
+
+  /**
+   * Prune columns from a {@link Schema} using a projected fields.
+   *
+   * @param schema a Schema
+   * @param projectedFields projected fields from Flink
+   * @return a Schema corresponding to the Flink projection
+   * @throws IllegalArgumentException if the Flink type does not match the Schema
+   */
+  public static Schema pruneWithoutReordering(Schema schema, List<String> projectedFields) {
+    if (projectedFields == null) {
+      return schema;
+    }
+
+    Map<String, Integer> indexByName = TypeUtil.indexByName(schema.asStruct());
+    Set<Integer> projectedIds = projectedFields.stream().map(indexByName::get).collect(Collectors.toSet());
+    return TypeUtil.select(schema, projectedIds);
```

Comment:
Continue with the question from [here](https://github.com/apache/iceberg/pull/1293#discussion_r469938063). If we could produce a  ordered & projected schema in this method (Saying if this method is `pruneWithReordering`), then seems we don't have to convert the read RowData to the correct order [here](https://github.com/apache/iceberg/pull/1346/files#diff-9e6ac35840fe0e9f8bacbe12c574c4eaR64) ? 

I'd prefer to use the correct projected schema to read the target RowData if possible, rather than reading RowData in a disordered schema and then order them in an iterator transformation.  Because this is in the critical read path and an extra RowData transformation will cost more resources , also make the codes hard to follow.

Reply:
> I'd prefer to use the correct projected schema to read the target RowData if possible

This is what I want too, but I'm afraid the current format readers do not have this capability. You can take a look to `AvroSchemaWithTypeVisitor`, the readers order is according to file schema instead of Flink projected/expected schema.

> Because this is in the critical read path and an extra RowData transformation will cost more resources

The performance is OK, because we just use a lazy projection in `ProjectionRowData`, Unnecessary projections are omitted.

Reply:
cc @rdblue , Mind to take a look this ? Thanks.

Reply:
You should be able to use the expected/projection schema. All readers should reorder columns to produce the requested column order.

`AvroSchemaWithTypeVisitor` traverses the file schema to create the reader structure, but that's because fields in Avro must be read in the file's order. But when that reader adds data columns to records, the values are put in the correct order because the [`ResolvingDecoder` returns the correct position](https://github.com/apache/iceberg/blob/master/core/src/main/java/org/apache/iceberg/avro/ValueReaders.java#L645-L648) in the projection schema.

Reply:
Thanks @openinx and @rdblue ! I'm very happy to be able to solve my confusion. I will do it.

Reply:
It doesn't matter if we change the order of the schema, as long as the ID doesn't change.